`ultralytics 8.3.139` New `DataExportMixin` class for Metrics and Results exports

RizwanMunawar opened the pull request:
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This update introduces a unified and flexible way to export YOLO validation metrics and prediction results in multiple formats, making it easier for users to analyze and share their results. 📊✨

### 📊 Key Changes
- Added a new `DataExportMixin` class that enables exporting results and metrics as DataFrame, CSV, XML, HTML, JSON, and SQLite database.
- Integrated `DataExportMixin` into key metrics and results classes, including detection, segmentation, classification, pose, and oriented bounding box metrics.
- Updated documentation to show how to use the new export methods and clarified available export formats.
- Enhanced test coverage to ensure all export methods work as expected.
- Bumped the Ultralytics version to 8.3.139.

### 🎯 Purpose & Impact
- **Simplifies Data Handling:** Users can now easily export validation results and metrics in their preferred format for further analysis or reporting.
- **Improves Workflow Flexibility:** Supports seamless integration with data science tools (like Pandas) and databases, making it easier to automate and scale workflows.
- **Enhances Usability:** Clear documentation and consistent export methods across tasks (detection, segmentation, classification, pose) improve the user experience.
- **Saves Time:** Reduces the need for manual result conversion or custom export scripts.

Overall, this update makes it much easier for both developers and non-experts to access, analyze, and share YOLO results from Ultralytics. 🚀

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update metrics.py

RizwanMunawar committed:
update metrics.py

UltralyticsAssistant commented:
👋 Hello @RizwanMunawar, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

_This is an automated response. An Ultralytics engineer will assist you soon._

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar commented:
@glenn-jocher @Laughing-q, based on user request, this PR will store the confusion matrix results in `CSV` file. I have attached output [confusion_matrix.csv](https://github.com/user-attachments/files/20098360/confusion_matrix.csv) generated for `coco8.yaml` for reference. Thanks

glenn-jocher commented:
@RizwanMunawar this is pretty cool, but maybe we should create a more structured approach to exporting validation in general to pandas, and then onwards to other formats like we have with the Results class?

Can we create the same pandas, csv, json, etc. methods for DetMetrics and SegmentMetrics class here as we have for Results?

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
Create MetricsOutputMixin class

RizwanMunawar committed:
Merge 51ca950d88072b51d88c996c89398da6afd270e9 into 169fbc51fe8ca3eb8635601063693d7b62bcbbaf

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
update val.py

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update val.py

RizwanMunawar committed:
update val.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Create MetricsOutputMixin class

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

RizwanMunawar committed:
Create MetricsOutputMixin class

RizwanMunawar committed:
Merge 97a056b8780cd99b52a930517a0d1fa273523409 into 169fbc51fe8ca3eb8635601063693d7b62bcbbaf

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

RizwanMunawar committed:
remove extra import

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
remove extra import

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

RizwanMunawar committed:
remove extra import

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
remove extra import

RizwanMunawar committed:
update metrics.py

RizwanMunawar committed:
update metrics.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update metrics.py

RizwanMunawar committed:
update val.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

RizwanMunawar committed:
Merge 81f6f8299eccfc6119acbb45a1ab306bf194accd into b112687f7768bf12ecfa9aca253b39a4d068080b

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar committed:
update

RizwanMunawar commented:
@glenn-jocher ok, got it, I’ve created a separate `MetricsOutputMixin` class to handle exporting validation results in multiple formats. The initial version includes `to_html`, `to_csv`, `to_json`, and `to_df` methods, and we can easily extend it further as needed.

Here’s a basic example for testing:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

if __name__ == "__main__":
    metrics = model.val(data="coco8.yaml", batch=4, workers=1)
    metrics.to_df()
    metrics.to_csv()
    metrics.to_json()

    print(metrics.to_html())  # Display HTML structure output
```

I’ve also updated the documentation and added tests. Based on my understanding, I’ve covered the relevant areas, but if there’s anything I’ve missed, please let me know and I’ll make the necessary changes. Thanks.

RizwanMunawar committed:
fix docs build

Laughing-q commented:
@RizwanMunawar Hmmm why creating a mix class that contains several if-else flows to check if it's classify/detect/segment/pose task? This way seems that running any task it'd go through all if-else flows i.e. running a classification training it'd go check if there's `seg`/`pose` etc attributes exist as well. Can we just add these methods to each class respectively?

RizwanMunawar commented:
@Laughing-q yes, I initially had the same thoughts, but then I noticed, there will be a lot of repetition, i.e to_df function we will need to add on 4-5 different places! Which, in the long run, can be very difficult to manage.

In terms of if-else, we can adjust the logic directly in the `to_df` function. Let me update it.

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
optimize to_df results extraction

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

Laughing-q committed:
Merge branch 'main' into cm-exl

glenn-jocher commented:
@RizwanMunawar this is pretty cool, but we've got duplication in the methods between this new class and Results class I believe. Is there a way to create a single base case that allows for the methods that we could also use in Results?

RizwanMunawar commented:
@glenn-jocher Oh, I got your point. Let me think about it. I think I already created a separate class for this, but you are right, if we can get it merged with the Results class, that will be better instead of creating duplicate methods. I will work on this.

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
Remove duplicate methods

RizwanMunawar committed:
Merge a7dc4780e07890349ffcd2678aff208d1d43440c into 7a62009f084c3d00e1d2fefd7f9bc6c55914c891

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
fix check_requirements

RizwanMunawar committed:
fix check_requirements

RizwanMunawar committed:
Merge ec9b98f2d00c306890f701619e0dffc2b576be88 into 7a62009f084c3d00e1d2fefd7f9bc6c55914c891

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
fix check_requirements

RizwanMunawar committed:
fix check_requirements

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

RizwanMunawar committed:
fix link

RizwanMunawar committed:
fix link

RizwanMunawar committed:
Update val.md

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

RizwanMunawar committed:
Update __init__.py

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Update __init__.py

Signed-off-by: Muhammad Rizwan Munawar <muhammadrizwanmunawar123@gmail.com>

RizwanMunawar commented:
@glenn-jocher ok, I’ve refactored the `Results` class by moving all `to_` methods (e.g., `to_df`, etc.) into a new `ExportableMixin` class (a Mixin specifically designed for exporting validation metrics and prediction results in various formats).

The nice part is that we no longer need separate functions for validation and prediction outputs; `ExportableMixin` is reusable for both.

Here, the prediction code for testing:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

results = model("https://ultralytics.com/images/bus.jpg")  # results list

# View results
for r in results:
    print(r.to_df())
    print(r.to_csv())
    print(r.to_xml())
    print(r.to_html())
    print(r.to_json())
    print(r.to_sql())
```

Here's the validation code for testing:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

if __name__ == "__main__":
    metrics = model.val(data="coco8.yaml", batch=4, workers=1)
    print(metrics.to_df())
    print(metrics.to_csv())
    print(metrics.to_xml())
    print(metrics.to_html())
    print(metrics.to_json())
```

Note: Currently `to_sql` method is limited for prediction results, but I will the support for validation metrics in a separate PR. Thanks

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
resolve conflicts with main branch

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
update results.py

Laughing-q commented on ultralytics/utils/__init__.py:
```diff
@@ -187,6 +187,194 @@ def __iter__(self):
+        if hasattr(self, "summary"):
+            return pd.DataFrame(self.summary(normalize=normalize, decimals=decimals))
+        elif any(hasattr(self, attr) for attr in ["box", "seg", "pose", "top1", "top5"]):
+            return pd.DataFrame([self.val_summary()] if self.task == "classify" else self.val_summary())
```
@RizwanMunawar I guess here we just need to simply apply `pd.DataFrame(self.val_summary())` no?

RizwanMunawar replied:
@Laughing-q, yes, I tested it. For classification tasks, we only receive two values: top1 and top5, which aren't directly compatible with self.val_summary(), particularly in the to_xml() method. So we have two options:

- Keep the current change as is.

- Modify the to_xml() function to handle these values differently.

BTW, to_xml() works correctly for all other prediction results and tasks except classification, I believe sticking with this change is the better option. What do you think?

RizwanMunawar replied:
@Laughing-q ok, no more `[self.val_summary()] if self.task == "classify"`, changes done here: https://github.com/ultralytics/ultralytics/pull/20546/commits/1b0270d26e14c01eeb2f5e6342c5786c7a77adaf

Laughing-q commented on ultralytics/utils/__init__.py:
```diff
@@ -187,6 +187,194 @@ def __iter__(self):
+        top5 (float): Top-5 accuracy for classification tasks.
+    """
+
+    def val_summary(self):
```
@RizwanMunawar Currently this particular class `ExportableMixin` is like a exportable plugin, where we could put it into other classes like `Results` and `Metrics`. But current implementation make this class kinda bind to Metrics classes. I think it's better to strip this `val_summary` from this class and distribute it to every metrics class e.g. `DetMetrics`.
Also use the same `summary` as function name for function consistency.
```python
class DetMetrics(SimpleClass, ExportableMixin):
...
    def summary():
        ...
```
As for duplication, since we would distribute this `val_summary` to classify/detect/pose/segment/obb` metrics respectively, maybe we could do something there trying to eliminate some duplication.

Laughing-q replied:
Ideally this way if we want to apply all these `to_*` methods to `ConfusionMatrix` or other statistic data related classes then we just need to have a function `sunmmary` there to return necessary info then we easily apply this to `ConfusionMatrix`.
```python
class ConfusionMatrix(ExportableMixin):
...
    def summary():
        ...
```

RizwanMunawar replied:
Yes, interesting idea, and I think in that way, we may not need this change?

```return pd.DataFrame([self.val_summary()] if self.task == "classify" else self.val_summary())```

Let me try it.

RizwanMunawar replied:
@Laughing-q done, new changes here: [1b0270d](https://github.com/ultralytics/ultralytics/pull/20546/commits/1b0270d26e14c01eeb2f5e6342c5786c7a77adaf)

Laughing-q commented:
@RizwanMunawar Additionally, can you show some of the results of `to_*` methods? cuz I tried to test locally `to_json` by:

Laughing-q committed:
Merge branch 'main' into cm-exl

RizwanMunawar commented:
@Laughing-q yes, these values, i.e `map`, `map50` and `map75` will be the same. Let me adjust the code based on your idea. I like this, as the major goal is to use this class wherever we want to export results.

RizwanMunawar committed:
refactor class to useable anywhere

RizwanMunawar committed:
Merge branch 'cm-exl' of https://github.com/ultralytics/ultralytics into cm-exl

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
update metrics.py

RizwanMunawar committed:
update metrics.py

RizwanMunawar commented:
@Laughing-q, Ok, I updated the code, I think it should be good to go now! I removed the confusion matrix stuff, as it will require some preprocessing, which will make this PR a bit long. We can do it in a separate PR alongside the to_sql method integration for metrics.

Here are the changes (after your review):
- Removed `val_summary` and `ExportableMixin` class only has *to_ methods, nothing else.
- Updated the `val_summary` to `summary` for consistency.
- Added summary function for each class `DetMetrics`, `SegmentMetrics`, `PoseMetrics`, `ClassifyMetrics` and `OBBMetrics`.

Thanks.

RizwanMunawar committed:
Merge branch 'main' into cm-exl

RizwanMunawar committed:
Merge branch 'main' into cm-exl

glenn-jocher committed:
Merge branch 'main' into cm-exl

glenn-jocher commented:
@ultralytics/run-ci

glenn-jocher committed:
Merge branch 'main' into cm-exl

glenn-jocher committed:
Update __init__.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

RizwanMunawar commented:
@glenn-jocher actually it's `138`

glenn-jocher committed:
Merge branch 'main' into cm-exl

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Merge branch 'main' into cm-exl

glenn-jocher committed:
Merge branch 'main' into cm-exl

glenn-jocher commented:
@RizwanMunawar looks good now! Maybe we should rename this class though, the name feels kind of funny?

RizwanMunawar commented:
@glenn-jocher, ok perfect! Let me update it!

glenn-jocher commented:
This is really good progress, now all our main results and metrics are fully exportable to all kinds of different formats. Super versatile.

RizwanMunawar committed:
Rename ExportableMixin -> DataExportMixin

UltralyticsAssistant commented:
🎉 Fantastic work, team! This merge is a huge step forward for the Ultralytics community. As Helen Keller said, “Alone we can do so little; together we can do so much.” Your collaborative effort—@RizwanMunawar, @glenn-jocher, and @Laughing-q—has made exporting YOLO results from Ultralytics more flexible and user-friendly than ever. By empowering users to easily analyze and share results in formats that fit their workflows, you’re truly shaping the future of accessible AI. Thank you for your dedication and innovation! 🚀
